astropy/nddata/mixins/ndarithmetic.py

```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
-        # should not be necessary, but also does no harm.
-        if self.unit is None and operand.unit is None:
-            if result_unit is dimensionless_unscaled:
-                result_unit = None
+        if handle_meta is None:
+            kwargs['meta'] = None
+        elif not handle_meta:
+            if len(self.meta) == 0:
+                kwargs['meta'] = operand.meta
             else:
-                raise ValueError("arithmetic result was not unitless even "
-                                 "though operands were unitless")
-        data = data.value
-        new_wcs = deepcopy(self.wcs)
-
-        # Call __class__ in case we are dealing with an inherited type
-        result = self.__class__(data, uncertainty=None,
-                                mask=None, wcs=new_wcs,
-                                meta=None, unit=result_unit)
-
-        # Prepare to scale uncertainty if it is needed
-        if operand.uncertainty:
-            operand_uncert_value = operand.uncertainty.array
-
-        # By this point the arithmetic has succeeded, so the input units were
-        # consistent with each other given the operation.
-        #
-        # If the operation is addition or subtraction then need to ensure that
-        # the uncertainty of the operand is the same units as the result
-        # (which will be the same as self.unit).
-
-        # The data ought to also be scaled in this case -- for addition of
-        # a StdDevUncertainty this isn't really necessary but other
-        # uncertainties when added/subtracted may depend on both the operand
-        # uncertainty and the operand data.
-
-        # Since the .unit.to methods create a copy, avoid the conversion
-        # unless it is necessary.
-        if (operation in [np.add, np.subtract] and
-                self.unit != operand.unit):
-            operand_data = operand.unit.to(self.unit, operand.data)
-            if operand.uncertainty:
-                operand_uncert_value = operand.unit.to(self.unit,
-                                                       operand_uncert_value)
+                kwargs['meta'] = self.meta
         else:
-            operand_data = operand.data
+            kwargs['meta'] = self._arithmetic_meta(operation, operand,
+                                        meta_kwds_operate, meta_kwds_set)
+
+        # Wrap the individual results into a new instance of the same class.
+        return result, kwargs
+
+    def _arithmetic_data(self, operation, operand):
+        """
+        Calculate the resulting data
 
-        if operand.uncertainty:
-            # Create a copy here in case this is returned as the uncertainty
-            # of the result.
-            operand_uncertainty = \
-                operand.uncertainty.__class__(operand_uncert_value, copy=True)
+        Parameters
+        ----------
+        operation: `str`
+            see `NDArithmeticMixin.arithmetic` parameter description.
```

Comment:
The function works well for the ``data`` and ``meta`` but for ``uncertainty`` I have to do something like
```
if operator_func is np.add:
    propagate_func = self.propagate_add
elif ...
```
The old version of ``NDArithmetic`` kept ``operation`` (the string describing the operation), ``operator`` (the function that is called for operating on the data) and the ``propagate_func`` (the function that is called to propagate the uncertainties). Since we only need one such parameter because we can deduce the others from that one I randomly chose the ``operation``-string. But I can change that, makes the code a bit shorter for the operation on ``data`` and ``meta``.